lib/wibox/widget/base.lua

```diff
@@ -321,6 +330,133 @@ function base.place_widget_at(widget, x, y, width, height)
+
+    -- Create layouts based on metatable __call
+    local l = layout.is_widget and layout or layout()
+
+    -- Get the number of children widgets (including nil widgets)
+    local max, attributes, widgets = parse_table(content, l.allow_empty_widget)
+
+    -- Get the optional identifier to create a virtual widget tree to place
+    -- in an "access table" to be able to retrieve the widget
+    local id = attributes.id
+
+    -- Clear the internal attributes
+    attributes.id, attributes.layout, attributes.widget = nil, nil, nil
+
+    for k = 1, max do
+        -- ipairs cannot be used on sparse tables
+        local v, id2, e = widgets[k], id, nil
+        if v then
+            -- It is another declarative container, parse it
+            if not v.is_widget then
+                e, id2 = drill(ids, v)
+                widgets[k] = e
+            end
+
+            -- Place the widget in the access table
+            if id2 then
+                l  [id2] = e
+                ids[id2] = e
+                id2      = nil
+            end
+        end
+    end
+
+    -- Drawable don't have any set_children and it would make no sense, this
+    -- make an exeption for them
+    local f = l.set_children or l.set_widget
+    f(l, widgets)
+
+    -- Set layouts attributes
+    for attr, val in pairs(attributes) do
+        if l["set_"..attr] then
+            l["set_"..attr](l, val)
+        elseif type(l[attr]) == "function" then
+            l[attr](l, val)
+        else
+            l[attr] = val
+        end
+    end
+
+    return l, id
+end
+
+--- Set a declarative widget hierarchy description
+-- @param args An array containing the widgets disposition
+--
+-- Here is a simple example of a horizontal layout with 3 widgets, including one
+-- with a background and one that will only be displayed on screen 1
+--     mywibox:set_widgets {
+--         s == 1 and my_first_widget, -- Only display on screen 1
+--         my_second_widget,
+--         { -- Add a background color/pattern for my_third_widget
+--               my_third_widget,
+--               layout = wibox.widget.background(nil,beautiful.bg_focus),
+--         },
+--         layout = wibox.layout.fixed.horizontal,
+--     }
+--
+-- The `layout` array member can either be the name of the layout or a layout
+-- object itself. This allow additional parameters be passed to the layout. It
+-- is also possible to declare the layout outside of the declarative array and
+-- simply include it:
+--
+-- @usage local mylayout = wibox.widget.background()
+--    mylayout:set_bg("#ff0000")
+--    
+--    mywibox:set_widgets {
+--        mywidget,
+--        layout = mylayout
+--    }
+--
+function base.widget:setup(args)
+    local f,ids = self.set_widget or self.add or self.set_first,{}
+    local w, id = drill(ids, args)
+    f(self,w)
+    if id then
+        self[id] = w
+    end
+    self.by_id = ids
```

Comment:
Could this be called `widget_by_id`? This also sets a random field on a random widget, but since the key is constant, I guess I can live with that.